url8012c8d031_dmbb966_BBMS_tgz-pJ8-testbed_JNEATtestJ8/src/gui/GUIMouse.java

```diff
@@ -0,0 +1,104 @@
+package gui;
+
+import hex.Hex;
+import hex.HexOff;
+import hex.VaporEnum;
+import bbms.GlobalFuncs;
+import terrain.TerrainEnum;
+import unit.SideEnum;
+
+public class GUIMouse {
+
+	/**
+	 * On mouse click, paints the terrain type specified in parameter tE.
+	 * @param e
+	 * @param tE
+	 */
+	public static void SetPaintTerrain(java.awt.event.MouseEvent e, TerrainEnum tE) {
+		HexOff cursorHexOff = GUIMainDisp.pixelToHexOff(e.getX(), e.getY(), -GlobalFuncs.gui.GMD.defaultHexSize, -GlobalFuncs.gui.GMD.defaultHexSize);
+		Hex h = GlobalFuncs.scenMap.getHex(cursorHexOff.getX(), cursorHexOff.getY());		
+
+		PaintHex(h, tE);	
+	}
+	
+	/** As SetPaintTerrain, but does it for a blob of seven hexes instead of a single one.
+	 * 
+	 * @param e
+	 * @param tE
+	 */
+	public static void SetPaintTerrainBlob(java.awt.event.MouseEvent e, TerrainEnum tE) {
+		HexOff cursorHexOff = GUIMainDisp.pixelToHexOff(e.getX(), e.getY(), -GlobalFuncs.gui.GMD.defaultHexSize, -GlobalFuncs.gui.GMD.defaultHexSize);
+		Hex h = GlobalFuncs.scenMap.getHex(cursorHexOff.getX(), cursorHexOff.getY());		
+		
+		PaintHex(h, tE);
+		
+		// Now paints surrounding hexes
+		for (int i = 0; i < 6; i++) {
+			HexOff neighbor = HexOff.NeighborOff(cursorHexOff,  i);
+			h = GlobalFuncs.scenMap.getHex(neighbor.getX(), neighbor.getY());
+			PaintHex(h, tE);
+		}
+	
+	}
+	
+	public static void PaintHex(Hex h, TerrainEnum tE) {
+		if (h != null) {
+			if (h.tEnum == TerrainEnum.INVALID) return; 
+			
+			Hex replacement = new Hex(h.x, h.y, tE, 0);
+			replacement.CloneHexData(h);
+			
+			GlobalFuncs.scenMap.hexArray[h.x][h.y] = replacement;
+			GlobalFuncs.gui.repaint();
+		}
+		
+	}
+	
+	
+	
+	/**
+	 * On mouse click, adds a unit to the appropriate side.
+	 * @param e
+	 * @param unitType
+	 * @param side
+	 */
+	public static void SetPlaceUnit(java.awt.event.MouseEvent e, String unitType, SideEnum side) {		
+		HexOff cursorHexOff = GUIMainDisp.pixelToHexOff(e.getX(), e.getY(), -GlobalFuncs.gui.GMD.defaultHexSize, -GlobalFuncs.gui.GMD.defaultHexSize);
+		Hex h = GlobalFuncs.scenMap.getHex(cursorHexOff.getX(), cursorHexOff.getY());
+		GUI_NB.GCO("Adding " + unitType);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem:**
 This line of code lacks validation when processing input data through the following parameter: &#39;unitType&#39; (index: 1 | type: String). The parameter is exposed to external callers, because its enclosing class and method are publicly accessible. This means that upstream validation, if it exists, can be bypassed. Other validated parameters: &#39;side&#39;. The same parameter type is validated here for example: url8012c8d031_dmbb966_BBMS_tgz-pJ8-testbed_JNEATtestJ8/src/bbms/GlobalFuncs.java:380. Malicious, malformed, or unbounded inputs can cause unexpected runtime behavior or crashes, and can slow performance.

**Fix:**
 Add checks to ensure the validity of the parameter&#39;s value, such as testing it for nullness (for example, using the @NonNull annotation as [described in the Lombok library](https://projectlombok.org/features/NonNull)), emptiness, or equality. Or to prevent direct calls to it, [reduce the method&#39;s visibility](https://docs.oracle.com/javase/tutorial/java/javaOO/accesscontrol.html).

 Learn more about potential threats and guidance from the [Common Weakness Enumeration website](https://cwe.mitre.org/data/definitions/693.html) and the [OWASP Cheat Sheet series](https://cheatsheetseries.owasp.org/cheatsheets/Input_Validation_Cheat_Sheet.html).